public/app/plugins/datasource/loki/components/useLokiLabels.ts

```diff
@@ -0,0 +1,75 @@
+import { useState, useEffect } from 'react';
+import LokiLanguageProvider from 'app/plugins/datasource/loki/language_provider';
+import { CascaderOption } from 'app/plugins/datasource/loki/components/LokiQueryFieldForm';
+
+/**
+ *
+ * @param languageProvider
+ * @param languageProviderInitialised
+ * @param activeOption rc-cascader provided option used to fetch option's values that hasn't been loaded yet
+ *
+ * @description Fetches missing labels and enables labels refresh
+ */
+export const useLokiLabels = (
+  languageProvider: LokiLanguageProvider,
+  languageProviderInitialised: boolean,
+  activeOption: CascaderOption[]
+) => {
+  // State
+  const [logLabelOptions, setLogLabelOptions] = useState([]);
+  const [shouldTryRefreshLabels, setRefreshLabels] = useState(false);
+
+  // Async
+  const fetchOptionValues = async option => {
+    await languageProvider.fetchLabelValues(option);
+    setLogLabelOptions(languageProvider.logLabelOptions);
```

Comment:
This being async, shouldn't we handle possibility of component being unmounted here? I would assume hooks throw if used after unmount in the same way as `setState`.

Reply:
Good point